Bug 797824 - Inconsistent behavior in Loan Review

Genie-Liu opened the pull request:
fix the shift issue in loan review. More discussion is https://bugs.gnucash.org/show_bug.cgi?id=797824

Genie-Liu committed:
Bug 797824 - Inconsistent behavior in Loan Review

fix the shift issue in loan review

jralls commented:
Hmm, still not right. 6 Months of loan but only 5 in the  summary. Whole loan has the same problem.
<img width="659" alt="Screenshot 2025-01-24 at 15 16 28" src="https://github.com/user-attachments/assets/8a7ff038-c88e-44bf-b59e-0dd5ee77b4c1" />

<img width="296" alt="Screenshot 2025-01-24 at 15 15 57" src="https://github.com/user-attachments/assets/feeef6cb-171f-45c8-9f5c-5acc611c1da9" />

It's still an improvement: Before your change it only went for 4 months.

Genie-Liu commented:
Hmm, interesting. Thank you for the feedback, It seems that after the change, the first payments is correct now. But I have a question. In my current version, the number of loan payments is accurate, yet you mentioned there are only four? Could there be a difference in settings? Here are my parameters:

<img width="666" alt="Screenshot 2025-01-25 at 15 20 57" src="https://github.com/user-attachments/assets/c1b19403-759f-4ac1-9cd9-e8160cc0d809" />
<img width="722" alt="Screenshot 2025-01-25 at 15 21 14" src="https://github.com/user-attachments/assets/38ef68e7-09de-4972-ae8c-ed602aebe552" />
<img width="601" alt="Screenshot 2025-01-25 at 15 21 21" src="https://github.com/user-attachments/assets/11386073-f950-4181-a8d6-981458e23f5e" />

jralls commented:
Interesting. The difference is the start date: You're starting the loan and the payments on Feb. 1, I'm leaving the loan start date as today, the default, and changing the payment start to Feb 1. Your screenshot also shows the bad first payment meaning it's without your change. Here's what it looks like with the change:
<img width="265" alt="image" src="https://github.com/user-attachments/assets/65bb0098-f560-42f0-94c3-d0acbd3a26bb" />

So it's correct as long as the loan start and payment start are the same.

Genie-Liu commented:
I see. I think my fix has resolved the issue with the first incorrect payment.

The missing payment in your previous figure is due to the start date configuration. The current mechanism calculates the end date based on the start date and the number of periods. In your earlier setting, the start date was set to the default, let’s assume it was 01/25/2025, with a 6-month duration. The end date would then be calculated as 06/25/2025. However, if we start the payment from 02/01/2025 and aim for 6 payments, the last payment would fall on 07/01/2025, which is after the calculated end date and it will be filtered during calculating the review.

A propose solution is to calculate the payments review using the remain months. This also aligns with the final created schedules.

Genie-Liu commented:
If we aim to create a loan schedule starting in the middle of the loan term, we will still have issues. For example, suppose the loan begins on 12/01/2024 with a 6-month period, and we intend to start payments on 02/01/2025. 

The current result is incorrect cause the first payment is wrong and the last payment is missing:
<img width="391" alt="image" src="https://github.com/user-attachments/assets/18830065-d117-4e1a-b739-550276caa78b" />

If we apply this change the fixed version should produce the payments starting at 02/01/2025 and end at 05/01/2025. The payments' date is correct. But the interest is still not the right one.

jralls commented:
>The missing payment in your previous figure is due to the start date configuration. The current mechanism calculates the end date based on the start date and the number of periods.

Fair enough, but a partial payment on the first day of the loan is called a down payment and technically isn't part of the loan. For house mortgages the first payment is usually near the beginning of the following month; auto loans may defer payments for up to a year.

> If we aim to create a loan schedule starting in the middle of the loan term, we will still have issues. For example, suppose the loan begins on 12/01/2024 with a 6-month period, and we intend to start payments on 02/01/2025.

>If we apply this change the fixed version should produce the payments starting at 02/01/2025 and end at 05/01/2025. The payments' date is correct. But the interest is still not the right one.

A 6-month loan beginning on 12/1 is due on 5/31. Strictly speaking if we start monthly payments on 2/1 then we would expect to have 5 payments of roughly $210 with the last one on 5/31.  But we have a problem in the Months Remaining box: If we leave it to itself it changes the number to 4 and assumes that we've already made two payments; if we push it to 6 it decides that we have until 7/1 to pay. Fixing that's out of scope for the bug and is deeper in the logic of the assistant. I'm satisfied that your current change fixes the date issues sufficiently to satisfy the bug. I'll do a code review tomorrow (it's already past my bedtime), but assuming it all looks good I'm willing to merge this as-is and I'll leave it up to you whether to work on getting the interest calculations right on a separate PR.

Genie-Liu commented:
> Fixing that's out of scope for the bug and is deeper in the logic of the assistant. I'm satisfied that your current change fixes the date issues sufficiently to satisfy the bug.

Yeah, I agree. The generation logic need more discussion. I will revert the repayments generation code that based on remain month. 

> I'll leave it up to you whether to work on getting the interest calculations right on a separate PR.

The interest fix is also included in this pr. It depends on the remain month. Maybe you can help take a look.

jralls commented:
You locked in on monthly payments but the payments option has several options. For example setting payments to semi-monthly produces:
<img width="762" alt="Screenshot 2025-01-26 at 08 57 43" src="https://github.com/user-attachments/assets/45a0b42a-0af1-4f87-8067-cb6539d0eaaf" />

<img width="781" alt="Screenshot 2025-01-26 at 08 57 56" src="https://github.com/user-attachments/assets/7a613065-fe72-450b-b998-f123e918bc30" />
<img width="287" alt="Screenshot 2025-01-26 at 08 58 23" src="https://github.com/user-attachments/assets/f2433304-b1dd-473b-a065-a9f7357d2123" />

**Edit:**
~~Without your changes the result is more correct though it of course ignores the start of payments date:~~
Without your changes the result has the right number of payments but doesn't adjust the payment amount so it overpays by 2x:

<img width="343" alt="Screenshot 2025-01-26 at 09 04 12" src="https://github.com/user-attachments/assets/c4d5c4db-c4b7-4f60-95d5-edd10506acbb" />

jralls commented:
Notice that the formula (labeled Amount) is unaffected by the payments schedule. That explains a lot of the problem. I wondered if it had been noticed and found [a lot of bugs](https://bugs.gnucash.org/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&bug_status=NEEDINFO&bug_status=REOPENED&component=Scheduled%20Transactions&list_id=24303&longdesc=mortgage&longdesc_type=allwordssubstr&order=Importance&query_format=advanced) documenting problems with it.

Genie-Liu commented:
It appears there are several issues that still require to fix. I believe there's at least four issues need to be addressed.

1. The initial payment date is incorrect.. This is what I plan to address. 
2. If the start date does not align with the repayment date, the final payment is omitted. This issue arises when the start date is left as default, . The root cause is how we calculate the end date. (If you set the length to be 1 year, you will observe that no payments are generated in the review at all)
3. The calculations for interest and principal are inaccurate when the payment date does not align with the loan's start date. This issue arises due to the review calculation beginning with an index of 1.
4. The current system does not support non-monthly repayment frequencies. 

This PR will focus on addressing issues 1 and 2, the monthly repayment schedule as it is the most commonly used. 
Issues 3 and 4 will be tackled in a separate PR.

Genie-Liu commented:
@jralls Please help test if the first and second issue are resolved after the new change.

The expected behavior should be: 
1. Setting loan start date to default like 01/27/2025 with length 6 months, and repayment start date at 02/01/2025. There should be 6 repayments in the review. 
2. Changing length to 1 year will also fix the no repayments issue.

jralls commented:
@Genie-Liu the date part seems to work as expected now. Please squash out the extra commits.

A note about item 3: The interest is wrong *unless* the the loan start date is exactly one month before the start of payments. The GnuCash [pmt function](https://github.com/Gnucash/gnucash/blob/55465e88ed33ba3ecb9a08cc4c6decb52182372e/bindings/guile/fin.scm#L48) works exactly like the spreadsheet function of the same name, and the last argument is a boolean `Type` where `false`/`0` means to calculste the interest at the *end* of the payment period while a `true` (anything not `0`) means to calculate it at the *beginning* of the period. As the Amount formula shows--and you can verify this easily with a spreadsheet--we're using `0` for `Type`.

Genie-Liu commented:
Commits squashed. Thanks for the tips.

jralls commented:
Thanks!
